metric_tank/cassandra.go

```diff
@@ -116,33 +116,38 @@ func searchCassandra(key string, start, end uint32) ([]Iter, error) {
 	query := func(month uint32, q string, p ...interface{}) {
 		wg.Add(1)
 		go func() {
-			//		if len(p) == 3 {
-			//		log.Println("querying cassandra for", q, p[0], TS(p[1]), TS(p[2]))
-			//		} else if len(p) == 2 {
-			//		log.Println("querying cassandra for", q, p[0], TS(p[1]))
-			//		} else {
-			//		log.Println("querying cassandra for", q, p)
-			//		}
 			results <- outcome{month, cSession.Query(q, p...).Iter()}
 			wg.Done()
 		}()
 	}
 
-	start_month := start - (start % month_sec)       // starting row has to be at, or before, requested start
-	end_month := (end - 1) - ((end - 1) % month_sec) // ending row has to be include the last point we might need
+	// unfortunately in the database we only have the t0's of all chunks.
+	// this means we can easily make sure to include the correct last chunk (just query for a t0 < end, the last chunk will contain the last needed data)
+	// but it becomes hard to find which should be the first chunk to include. we can't query for start <= t0 because than we will miss some data,
+	// we effectively need all chunks with a t0 >= start, as well as the last chunk with a t0 <= start which btw, could also be in the previous row.
+	// and also we can't assume we know the chunkSpan so we can't just calculate the t0 >= start - <some-predefined-number> because chunkSpans may change over time.
+	// but it seems cumbersome+slow to first query something like `where ts <= start order by ts desc limit 1` to check if that yielded a chunk, and if not, go to
+	// the previous row. and by the way, it doesn't even seem possible to execute such query in cassandra anyway.
+	// i wish we could just add another property to each ts-data combo to mark the "last-ts" contained within the chunk. that would make things a lot easier.
+	// but since i'm told that's impossible and given all of the above we can just assume that a chunkspan will never be more than 12h (TODO: might want to make this configurable)
+	// and just query 12 hours back and filter out the data.  this comes with a bunch of io overhead but what you gonna do...
+
+	conservative_start := start - 12*60*60
+	start_month := conservative_start - (conservative_start % month_sec) // starting row has to be at, or before, requested start.
```

Comment:
I really dont like this idea.  Having to fetch an additional 12hours of data is excessive.

This can all be solved by using the query originally discussed.

`("select ts, data from metric where key=? AND ts < ? Limit 1", row_key, start)`    Chunks are stored in descending order,as per the schema on line 28